### PR TITLE
Kinesis Connector

### PR DESCRIPTION
**Pull request by npawar**
## Description
Adding Kinesis Connector for Pinot realtime ingestion https://github.com/apache/incubator-pinot/issues/5648

Added a `KinesisConsumerFactory` to provide the Kinesis implementation.

Each kinesis consumer consumes from a shard. An abstraction has been introduced for a group of partition/shards, and each consumer will now be responsible for a PartitionGroup instead of just a partitionId (though in first iteration, the PartitionGroup will only contain 1 partition/shard).

Kafka stream should function as is, with no changes needed.

Sample table config:
```
{
  "tableName": "kinesisTest",
  "tableType": "REALTIME",
  "segmentsConfig": {
    "timeColumnName": "DaysSinceEpoch",
    "schemaName": "airlineStats",
    "replicasPerPartition": "1"
  },
  "tenants": {},
  "tableIndexConfig": {
    "loadMode": "MMAP",
    "streamConfigs": {
      "streamType": "kinesis",
      "aws-region": "<region>",
      "max-records-to-fetch": "20",
      "shard-iterator-type": "AFTER_SEQUENCE_NUMBER",
      "stream.kinesis.consumer.type": "lowlevel",
      "stream.kinesis.topic.name": "kinesis-test",
      "stream.kinesis.fetch.timeout.millis": "10000",
      "stream.kinesis.decoder.class.name": "org.apache.pinot.plugin.stream.kafka.KafkaJSONMessageDecoder",
      "stream.kinesis.consumer.factory.class.name": "org.apache.pinot.plugin.stream.kinesis.KinesisConsumerFactory",
      "realtime.segment.flush.threshold.size": "10000",
      "realtime.segment.flush.threshold.time": "1h"
    }
  },
  "metadata": {
    "customConfigs": {}
  }
}
```

Some pending items:
1. Integration test - not able to find an easy way to create a test Kinesis. @KKcorps is looking into this.
2. Kinesis has shardIds (equivalent to Kafka partition id). These have the format `shardId-000000000001`. We are using the numeric suffix as the "partitionId" in Pinot. But our Pinot partitionId is int, and with Kinesis, we would need long. This change will be done in a followup, since changing this was touching a lot of files and exploding the scope of the review.
3. It would be nice to add a subclass to KafkaJSONDecoder as "JSONDecoder" so that the Kinesis table config doesn't have to refer to the "Kafka" decoder.

## Release Notes
Kinesis connector

## Documentation
Coming very soon

### CODE REVIEW

**Comment by snleee** [comment]
@npawar Can you add the design doc to this commit message? I remember that there was a doc for interfaces.

**Comment by npawar** [comment]
> @npawar Can you add the design doc to this commit message? I remember that there was a doc for interfaces.

linked to the Kinesis Issue, which has the design doc

File: pinot-common/src/main/java/org/apache/pinot/common/utils/SegmentName.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public String getGroupId() {
     throw new RuntimeException("No groupId in " + getSegmentName());
   }
 
-  public int getPartitionId() {
+  public int getPartitionGroupId() {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/SegmentName.java:66]: nit: Please change the exception message as well

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/LLCRealtimeSegmentZKMetadata.java [hunk 2]
```diff
@@ -87,11 +87,6 @@ public void setDownloadUrl(String downloadUrl) {
   public ZNRecord toZNRecord() {
     ZNRecord znRecord = super.toZNRecord();
     znRecord.setSimpleField(START_OFFSET, _startOffset);
-    if (_endOffset == null) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/LLCRealtimeSegmentZKMetadata.java:90]: Please flag the commit for release notes and add in checkin comments that in order for people to use this release, all components should have been upgraded to the previous release (or, 0.5.0, not sure. please verify). thanks

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 3]
```diff
@@ -449,13 +497,40 @@ private void commitSegmentMetadataInternal(String realtimeTableName,
...
+    if (newPartitionGroupSet.contains(committingSegmentPartitionGroupId)) {
+      LLCSegmentName newLLCSegment = new LLCSegmentName(rawTableName, committingSegmentPartitionGroupId,
+          committingLLCSegment.getSequenceNumber() + 1, newSegmentCreationTimeMs);
+      createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegment, newSegmentCreationTimeMs,
+          committingSegmentDescriptor, committingSegmentZKMetadata, instancePartitions, numPartitions, numReplicas);
+      newConsumingSegmentName = newLLCSegment.getSegmentName();
+    }
+
+    // TODO: Also, create new partition groups here (instead of waiting for the Validation Manager)
+    //  Cannot do it at the moment, because of the timestamp suffix on the segment name.
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:532]: I fixed that by adding a check in the idealstate. Only one partition group and seq number combination allowed.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 4]
```diff
@@ -149,22 +151,22 @@ public boolean isFinal() {
   public class SegmentBuildDescriptor {
     final File _segmentTarFile;
     final Map<String, File> _metadataFileMap;
-    final StreamPartitionMsgOffset _offset;
+    final Checkpoint _offset;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:154]: Why rename this to Checkpoint?

**Review by mcvsubbu — COMMENTED** [review]
Partial. will continue

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java (hunk 4 above)
Comment (KKcorps) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:154]: `StreamPartitionMsgOffset` now extends `Checkpoint` so base class was kept here. It is not compulsory now to implement MsgOffset

File: pinot-spi/src/main/java/org/apache/pinot/spi/stream/Checkpoint.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.stream;
+
+/**
+ * Keeps track of the consumption for a PartitionGroup
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/stream/Checkpoint.java:22]: I still don't understand why we need this class. Why not just extend StreamPartitionOffset interface to include a fromString() method? (we already have a toString()).

I am also not opposed to adding two new methods serialize() and deserialize() if that somehow (I don't see how) makes things look/feel better. We can deprecate the existing toString() method.

Comment (KKcorps) [reply pinot-spi/src/main/java/org/apache/pinot/spi/stream/Checkpoint.java:22]: @npawar I also feel the same now that we should stick to a single class. Was there a specific motivation behind this change? If not, I'll make the changes suggested by @mcvsubbu

Comment (npawar) [reply pinot-spi/src/main/java/org/apache/pinot/spi/stream/Checkpoint.java:22]: Alright, lets not add Checkpoint.

**Pull request closed without merging**